src/primval.rs

```diff
@@ -14,6 +14,8 @@ pub enum PrimVal {
 
     Ptr(Pointer),
     FnPtr(Pointer),
+    VtablePtr(Pointer, Pointer),
+    SlicePtr(Pointer, u64),
```

Comment:
I'm not sure I can make a better argument without re-reading the code in full and weighing the pros/cons again. I'm inclined to merge and re-visit the question later.